src/coreclr/src/jit/rationalize.cpp

```diff
@@ -768,6 +768,32 @@ Compiler::fgWalkResult Rationalizer::RewriteNode(GenTree** useEdge, Compiler::Ge
         break;
 #endif // FEATURE_SIMD
 
+#ifdef FEATURE_HW_INTRINSICS
+        case GT_HWINTRINSIC:
+        {
+            GenTreeHWIntrinsic* hwIntrinsicNode = node->AsHWIntrinsic();
+
+            if (!hwIntrinsicNode->isSIMD())
+            {
+                break;
+            }
+
+            noway_assert(comp->supportSIMDTypes());
+
+            // TODO-1stClassStructs: This should be handled more generally for enregistered or promoted
+            // structs that are passed or returned in a different register type than their enregistered
+            // type(s).
+            if ((hwIntrinsicNode->gtType == TYP_I_IMPL) && (hwIntrinsicNode->gtSIMDSize == TARGET_POINTER_SIZE))
+            {
+                // This happens when it is consumed by a GT_RET_EXPR.
+                // It can only be a Vector2f or Vector2i.
+                assert(genTypeSize(hwIntrinsicNode->gtSIMDBaseType) == 4);
+                hwIntrinsicNode->gtType = TYP_SIMD8;
+            }
```

Comment:
The `GT_SIMD` path just above has some additional logic which checks if the operand types are `TYP_STRUCT` and changes them to `node->gtType` if they are.

Such logic is assuming that all operands are the same size, which isn't the case for HWIntrinsics, and so I didn't add it here. It might be an optimization that is also needed eventually.